text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,854 @@
+-     unsafe data: Box<[MaybeUninit<u8>]>,
+      /// SAFETY: See [`BoxErased::data`].
++     unsafe data: ManuallyDrop<Box<[MaybeUninit<u8>]>>,
+      unsafe type_id: TypeId,
+  }
+```
+
+## When *Not* To Use Unsafe Fields
+
+### Relaxing a Language Invariant
+
+The `unsafe` modifier is appropriate only for denoting *library* safety invariants. It has no impact
+on *language* safety invariants, which must *never* be violated. This, for example, is an unsound
+API:
+
+```rust
+struct Zeroed<T> {
+    // SAFETY: The value of `zeroed` consists only of bytes initialized to `0`.
+    unsafe zeroed: T,
+}
+
+impl<T> Zeroed<T> {
+    pub fn zeroed() -> Self {
+        unsafe { Self { zeroed: core::mem::zeroed() }}
+    }
+}
+```
+
+...because `Zeroed::<NonZeroU8>::zeroed()` induces undefined behavior.
+
+### Denoting a Correctness Invariant
+
+A library *correctness* invariant is an invariant imposed by an API whose violation must not result
+in undefined behavior. In the below example, unsafe code may rely upon `alignment_pow`s invariant,
+but not `size`'s invariant:
+
+```rust
+struct Layout {
+    /// The size of a type.
+    ///
+    /// # Invariants
+    ///
+    /// For well-formed layouts, this value is less than `isize::MAX` and is a multiple of the alignment.
+    /// To accomodate incomplete layouts (i.e., those missing trailing padding), this is not a safety invariant.
+    pub size: usize,
+    /// The log₂(alignment) of a type.
+    ///
+    /// # Safety
+    ///
+    /// `alignment_pow` must be between 0 and 29.
+    pub unsafe alignment_pow: u8,
+}
+```
+
+The `unsafe` modifier should only be used on fields with *safety* invariants, not merely correctness
+invariants.
+
+We might also imagine a variant of the above example where `alignment_pow`, like `size`, doesn't
+carry a safety invariant. Ultimately, whether or not it makes sense for a field to be `unsafe` is a
+function of programmer preference and API requirements.
+
+## Complete Example
+
+The below example demonstrates how field safety support can be applied to build a practical
+abstraction with small safety boundaries
+([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=e8aa2af933f5bf4892d1be951062538d)):
+
+```rust
+#![deny(
+    unfulfilled_lint_expectations,
+    clippy::missing_safety_doc,
+    clippy::undocumented_unsafe_blocks,
+)]
+
+use std::{
+    cell::UnsafeCell,
+    ops::{Deref, DerefMut},
+    sync::Arc,
+};
+
+/// An `Arc` that provides exclusive access to its referent.
+///
+/// A `UniqueArc` may have any number of `KeepAlive` handles which ensure that
+/// the inner value is not dropped. These handles only control dropping, and do
+/// not provide read or write access to the value.
+pub struct UniqueArc<T: 'static> {
+    /// # Safety
+    ///
+    /// So long as `T` is owned by `UniqueArc`, `T` may not be accessed (read or
+    /// written) other than via this `UniqueArc`.
```

Comment:
When is `T` owned by `UniqueArc`? Without defining that, this invariant is insufficiently defined and I don't know how to make sense of it.

Reply:
I think this safety comment could be slightly reworded. IIUC the intention is that *so long as a `u: UniqueArc` instance exists*, then `u.arc`'s referent may not be accessed other than via `u`.

Reply:
Would it be correct to rephrase this as "While the `UniqueArc` exists, the value pointed by this `Arc` must not be accessed (read or written) through any other pointer"? Given the API, I imagine this could also just be strengthened to "this must be the only `Arc` pointing to that location".

Reply:
> Would it be correct to rephrase this as "While the `UniqueArc` exists, the value pointed by this `Arc` must not be accessed (read or written) through any other pointer"?

Yes, I think that's correct. I've made that edit: https://github.com/rust-lang/rfcs/pull/3458/commits/322c4a871f6ad0a272ef5151b702cd9a1932864c

> Given the API, I imagine this could also just be strengthened to "this must be the only `Arc` pointing to that location".

Not quite, since `KeepAlive` also holds an `Arc` to the same location.